controller/api/proxy/fallback_profile_listener.go

```diff
@@ -0,0 +1,73 @@
+package proxy
+
+import (
+	"sync"
+
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+)
+
+type fallbackProfileListener struct {
+	underlying profileUpdateListener
+	state      *sp.ServiceProfile
+	primary    *fallbackProfileListener
+	backup     *fallbackProfileListener
+	mutex      *sync.Mutex
+}
+
+// newFallbackProfileListener takes an underlying profileUpdateListener and
+// returns two profileUpdateListeners: a primary and a secondary.  Updates to
+// the primary and secondary will propagate to the underlying with updates to
+// the primary always taking priority.  If the value in the primary is cleared,
+// the value from the secondary is used.
+func newFallbackProfileListener(listener profileUpdateListener) (profileUpdateListener, profileUpdateListener) {
+	// Primary and secondary share a lock to ensure updates are atomic.
+	mutex := sync.Mutex{}
+
+	primary := fallbackProfileListener{
+		underlying: listener,
+		mutex:      &mutex,
+	}
+	secondary := fallbackProfileListener{
+		underlying: listener,
+		mutex:      &mutex,
+	}
+	primary.backup = &secondary
+	secondary.primary = &primary
+	return &primary, &secondary
```

Comment:
I'm struggling to keep keep track of which fields in `fallbackProfileListener` are relevant in which contexts, how about something like:
```go
type fallbackProfileListener struct {
	underlying profileUpdateListener
	primary    *spListener
	backup     *spListener
	mutex      *sync.Mutex
}

type spListener struct {
	state  *sp.ServiceProfile
	parent *fallbackProfileListener
}

func (f *spListener) Update(profile *sp.ServiceProfile) {
```
This structure feels slightly closer to what the data represents, as it implies:
- one copy of `underlying` and `mutex`
- `state` does not exist on `parent`

Reply:
Nice suggestion, thanks.  I also split primary and backup into separate types with separate Update implementations to make things more clear.